### PR TITLE
chore(httpPost): Adding http post method in httpProbe

### PR DESCRIPTION
**Pull request by ispeakc0de**
Signed-off-by: shubhamchaudhary <shubham.chaudhary@mayadata.io>

- Adding http post method inside the http probe
-  Adding criteria match for the response code for HTTP Get request

### ChaosEngine For HTTP GET

```yaml
probe:
      - name: http-get-request
        type: "httpProbe"
        httpProbe/inputs:
          url: "<http url>"
          method:
            get:
              criteria: "=="
              responseCode: "200"
        mode: "Edge"
        runProperties:
          probeTimeout: 10
          interval: 5
          retry: 1
          probePollingInterval: 2
```

### ChaosEngine For HTTP POST

```yaml
probe:
      - name: http-post-request
        type: "httpProbe"
        httpProbe/inputs:
          url: "<http-url>"
          method:
            post:
              bodyPath: "/mnt/body.yml"
              contentType: "application/json; charset=UTF-8"
              criteria: "=="
              responseCode: "200"
        mode: "Edge"
        runProperties:
          probeTimeout: 10
          interval: 5
          retry: 1
          probePollingInterval: 2
```

### CODE REVIEW

File: pkg/probe/httpprobe.go [hunk 1]
```diff
@@ -48,40 +53,142 @@ func TriggerHTTPProbe(probe v1alpha1.ProbeAttributes, resultDetails *types.Resul
...
-				}
-				client = &http.Client{Transport: transCfg}
+			code := strconv.Itoa(resp.StatusCode)
+
+			// comparing the response code with the expected criteria
+			if err = FirstValue(probe.HTTPProbeInputs.Method.Get.ResponseCode).
+				SecondValue(code).
+				Criteria(probe.HTTPProbeInputs.Method.Get.Criteria).
+				CompareInt(); err != nil {
+				log.Errorf("The %v http probe has been Failed, err: %v", probe.Name, err)
```
Comment (ksatchit) [inline pkg/probe/httpprobe.go:128]: can we specify explicitly thet post op has failed in log

File: pkg/probe/httpprobe.go [hunk 2]
```diff
@@ -48,40 +53,142 @@ func TriggerHTTPProbe(probe v1alpha1.ProbeAttributes, resultDetails *types.Resul
...
+	if httpBody.Body != "" {
+		return httpBody.Body, nil
+	}
+
+	var command string
+
+	if httpBody.BodyPath != "" {
+		command = "cat " + httpBody.BodyPath
+	} else {
+		return "", errors.Errorf("[Probe]: Any one of body or bodyPath is required")
```
Comment (uditgaurav) [inline pkg/probe/httpprobe.go:180]: go code can also be used?

**Comment by ispeakc0de** [comment]
```yaml
time="2021-02-15T11:51:01Z" level=info msg="Experiment Name: pod-delete"
time="2021-02-15T11:51:01Z" level=info msg="[PreReq]: Getting the ENV for the  experiment"
time="2021-02-15T11:51:01Z" level=info msg="[PreReq]: Updating the chaos result of pod-delete experiment (SOT)"
time="2021-02-15T11:51:01Z" level=info msg="The application information is as follows" Ramp Time=0 Namespace=shubham Label="run=nginx"
time="2021-02-15T11:51:01Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (pre-chaos)"
time="2021-02-15T11:51:01Z" level=info msg="[Status]: Checking whether application containers are in ready state"
time="2021-02-15T11:51:01Z" level=info msg="[Status]: The Container status are as follows" Readiness=true container=nginx Pod=nginx-78cb94bbd5-fc9nf
time="2021-02-15T11:51:03Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2021-02-15T11:51:03Z" level=info msg="[Status]: The status of Pods are as follows" Status=Running Pod=nginx-78cb94bbd5-fc9nf
time="2021-02-15T11:51:05Z" level=info msg="[Probe]: The http probe information is as follows" Mode=SOT Phase=PreChaos Name=http-post-probe URL="http://23.251.152.171:80/register" Run Properties="{10 5 1 2 0}"
time="2021-02-15T11:51:05Z" level=info msg="[Probe]: HTTP Post method informations" Name=http-post-probe URL="http://23.251.152.171:80/register" Body= BodyPath=/mnt/body.yaml ContentType="application/json; charset=UTF-8"
time="2021-02-15T11:51:10Z" level=info msg="[Probe]: http-post-probe probe has been Passed 😄 " ProbeStatus=Passed ProbeName=http-post-probe ProbeType=httpProbe ProbeInstance=PreChaos
time="2021-02-15T11:51:10Z" level=info msg="[Chaos]:Number of pods targeted: 1"
time="2021-02-15T11:51:10Z" level=info msg="Target pods list for chaos, [nginx-78cb94bbd5-fc9nf]"
time="2021-02-15T11:51:10Z" level=info msg="[Info]: Killing the following pods" PodName=nginx-78cb94bbd5-fc9nf
time="2021-02-15T11:51:10Z" level=info msg="[Wait]: Wait for the chaos interval 10s"
time="2021-02-15T11:51:20Z" level=info msg="[Status]: Verification for the recreation of application pod"
time="2021-02-15T11:51:20Z" level=info msg="[Status]: Checking whether application containers are in ready state"
time="2021-02-15T11:51:20Z" level=info msg="[Status]: The Container status are as follows" Pod=nginx-78cb94bbd5-t6p5p Readiness=true container=nginx
time="2021-02-15T11:51:22Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2021-02-15T11:51:22Z" level=info msg="[Status]: The status of Pods are as follows" Pod=nginx-78cb94bbd5-t6p5p Status=Running
time="2021-02-15T11:51:24Z" level=info msg="[Chaos]:Number of pods targeted: 1"
time="2021-02-15T11:51:24Z" level=info msg="Target pods list for chaos, [nginx-78cb94bbd5-t6p5p]"
time="2021-02-15T11:51:24Z" level=info msg="[Info]: Killing the following pods" PodName=nginx-78cb94bbd5-t6p5p
time="2021-02-15T11:51:24Z" level=info msg="[Wait]: Wait for the chaos interval 10s"
time="2021-02-15T11:51:34Z" level=info msg="[Status]: Verification for the recreation of application pod"
time="2021-02-15T11:51:34Z" level=info msg="[Status]: Checking whether application containers are in ready state"
time="2021-02-15T11:51:34Z" level=info msg="[Status]: The Container status are as follows" Pod=nginx-78cb94bbd5-wd9g7 Readiness=true container=nginx
time="2021-02-15T11:51:36Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2021-02-15T11:51:36Z" level=info msg="[Status]: The status of Pods are as follows" Pod=nginx-78cb94bbd5-wd9g7 Status=Running
time="2021-02-15T11:51:38Z" level=info msg="[Chaos]:Number of pods targeted: 1"
time="2021-02-15T11:51:38Z" level=info msg="Target pods list for chaos, [nginx-78cb94bbd5-wd9g7]"
time="2021-02-15T11:51:38Z" level=info msg="[Info]: Killing the following pods" PodName=nginx-78cb94bbd5-wd9g7
time="2021-02-15T11:51:38Z" level=info msg="[Wait]: Wait for the chaos interval 10s"
time="2021-02-15T11:51:48Z" level=info msg="[Status]: Verification for the recreation of application pod"
time="2021-02-15T11:51:48Z" level=info msg="[Status]: Checking whether application containers are in ready state"
time="2021-02-15T11:51:48Z" level=info msg="[Status]: The Container status are as follows" container=nginx Pod=nginx-78cb94bbd5-4h8c8 Readiness=true
time="2021-02-15T11:51:50Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2021-02-15T11:51:50Z" level=info msg="[Status]: The status of Pods are as follows" Pod=nginx-78cb94bbd5-4h8c8 Status=Running
time="2021-02-15T11:51:52Z" level=info msg="[Chaos]: Time is up for experiment: pod-delete"
time="2021-02-15T11:51:52Z" level=info msg="[Completion]: pod-delete chaos is done"
time="2021-02-15T11:51:52Z" level=info msg="[Confirmation]: The application pod has been deleted successfully"
time="2021-02-15T11:51:52Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (post-chaos)"
time="2021-02-15T11:51:52Z" level=info msg="[Status]: Checking whether application containers are in ready state"
time="2021-02-15T11:51:52Z" level=info msg="[Status]: The Container status are as follows" container=nginx Pod=nginx-78cb94bbd5-4h8c8 Readiness=true
time="2021-02-15T11:51:54Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2021-02-15T11:51:54Z" level=info msg="[Status]: The status of Pods are as follows" Pod=nginx-78cb94bbd5-4h8c8 Status=Running
time="2021-02-15T11:51:57Z" level=info msg="[The End]: Updating the chaos result of pod-delete experiment (EOT)"
```

File: pkg/probe/httpprobe.go (hunk 2 above)
Comment (ispeakc0de) [reply pkg/probe/httpprobe.go:180]: Didn't understand this. Can you please add more details?

**Review by ksatchit — APPROVED** [review]

File: pkg/probe/httpprobe.go (hunk 2 above)
Comment (uditgaurav) [reply pkg/probe/httpprobe.go:180]: I mean to say we can use `io` file reader in place of running the bash command (cat) here. It is ok if you want it to do it in the next PR as well.

**Review by uditgaurav — APPROVED** [review]

**Pull request merged**